develop/tutorials/articles/03-portlets/packaging-a-jsf-application.markdown

```diff
@@ -50,6 +46,51 @@ example anatomy structure is outlined below:
     - `faces-config.xml`
     - `web.xml`
 
+## Auto-Generated WAB [](id=auto-generated-wab)
+
+The source of an auto-generated JSF Portlet WAB follows the standard
+directory layout for WAR projects. Bnd directives for the manifest can be placed in the liferay-plugin-package.properties file.  For use of Declarative Services in your JSF portlet, a java class that extends [GenericFacesPortlet](http://myfaces.apache.org/portlet-bridge/2.0/api/apidocs/javax/portlet/faces/GenericFacesPortlet.html) may be included and annotated with `@Component` on the class and `@Reference` on a method to inject a service.
+
+- `src/`
+    - `main/`
+        - `java/`
+        - `webapp/`
+            - `WEB-INF/`
+                - `faces-config.xml`
+                - ~~`liferay-display.xml`~~ (no longer needed)
+                - `liferay-plugin-package.properties`
+                - ~~`liferay-portlet.xml`~~ (no longer needed)
+                - ~~`portlet.xml`~~ (no longer needed)
+                - `web.xml`
+
+**Benefits:**
+
+- Processed by the Liferay [WAB Generator](https://github.com/liferay/liferay-portal/tree/master/modules/apps/foundation/portal-osgi-web/portal-osgi-web-wab-generator), which automatically creates an OSGi-ready `META-INF/MANIFEST.MF`.
+- Processed by the Liferay auto-deploy process, which adds the `PluginContextListener` to the `WEB-INF/web.xml` descriptor when the WAB has not been pre-configured.
+- Can affect the content of `META-INF/MANIFEST.MF` by putting BND directives and OSGi headers into the `WEB-INF/liferay-plugin-package.properties` file.
```

Comment:
I just tried:
``-removeheaders=Created-By``

And got the following results:
```
$ bnd print `ls -tr liferay-portal-7.0.0-SNAPSHOT-jsf-2.2/osgi/wabs/com.liferay.faces.demo.primefaces.users.portlet-*.war | tail -2 | head -1` 2>>/dev/null | grep Created-By
Created-By                               1.8.0_25 (Oracle Corporation)           
Originally-Created-By                    Apache Maven 3.4.0-SNAPSHOT             
$ bnd print `ls -tr liferay-portal-7.0.0-SNAPSHOT-jsf-2.2/osgi/wabs/com.liferay.faces.demo.primefaces.users.portlet-*.war | tail -1 | head -1` 2>>/dev/null | grep Created-By
Originally-Created-By                    Apache Maven 3.4.0-SNAPSHOT
```
So, it looks like directives do get applied.  I am off to try and shoot myself in the foot now with:
``-removeheaders=Import-Package``

Cool, that results in the following error at deploy time:
```
20:44:03,670 ERROR [Framework Event Dispatcher: Equinox Container: 90361f22-8c29-0016-19de-b6e9c43dda49][com_liferay_portal_osgi_web_servlet_context_helper:97] FrameworkEvent ERROR 
org.osgi.framework.ServiceException: Exception in com.liferay.portal.osgi.web.servlet.context.helper.internal.ServletContextHelperRegistrationServiceFactory.getService()
	at org.eclipse.osgi.internal.serviceregistry.ServiceFactoryUse.factoryGetService(ServiceFactoryUse.java:222)
... snipped for brevity ...
```